src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,150 @@
+            MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager,
+            typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription2 = new LocalizableResourceString(
+            nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterDescription),
+            MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, 
+            typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        internal static DiagnosticDescriptor RuleReplaceOneParameter = new DiagnosticDescriptor(
+            RuleIdOneParameterId,
+            s_localizableTitle,
+            s_localizableMessageDefault,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        internal static DiagnosticDescriptor RuleReplaceTwoParameter = new DiagnosticDescriptor(
+            RuleIdTwoParameterId,
+            s_localizableTitle2,
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(this.OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext startContext)
+        {
+            var compilation = startContext.Compilation;
+
+            var stringSymbol = WellKnownTypes.String(compilation);
+
+            if (stringSymbol is null)
+                return;
+
+            var substring1ParameterMethod = stringSymbol.GetMembers(nameof(string.Substring)).OfType<IMethodSymbol>()
+                .SingleOrDefault(substring => substring.Parameters.Length == 1);
+
+            if (substring1ParameterMethod is null)
+                return;
+
+            var substring2ParameterMethod = stringSymbol.GetMembers(nameof(string.Substring)).OfType<IMethodSymbol>()
+                .SingleOrDefault(substring => substring.Parameters.Length == 2);
```

Comment:
needs to be FirstOrDefault.  SingleOrDefault will throw if there are two methods.

Reply:
That's correct, but if there are two methods I can't decide on the right one yet.
Now checking for Name, Length and types - As far as I know that should be unique for a given type (correct me if I'm wrong).

Reply:
Yes, needs to be changed to FirstOrDefault

---
In reply to: [226067368](https://github.com/dotnet/roslyn-analyzers/pull/1854#discussion_r226067368) [](ancestors = 226067368)